### DISCUSSION TITLE
Logging/Recording LHS and RHS of User Input

### DISCUSSION THREAD

**Discussion by exformation** [neovim/neovim#27936] [Q&A] [answered]
As a user types, I want to log what they type (the lhs of their mapping) and what was triggered (the rhs of their mapping). Ex:
```
gd           vim.lsp.buf.definition
j            j
ft           f{char}
<leader>b    :b#<cr>
...
```

I feel like this should exist but I'm having trouble finding anything in the Lua/Vim API.

vim.on_key() captures keypresses but they're after mappings are applied instead of before, and I haven't been able to find anything that shows what functions a user is calling, like capturing if they just called vim.lsp.buf.definition. debug.sethook() and debug.getinfo() seemed like the wrong rabbit hole for that. 

I could save and override all user mappings to include their original function and the logging:
```
set(modes, lh, log(lh,rh)<cr>rh, opts)
```
but I'm not sure if there will be a ton of edge-cases and it seems like a hacky solution.

**Comment by justinmk** [comment]
This is planned but we don't have it yet, afaik.

**Reply by exformation** [reply]
Do you know where I can track the status of this? I guess I'll do the work-around for now with plans to replace it.

**Reply by justinmk** [reply]
It's a key part of internals needed for "multicursor", so look for issues that mention that.

**Comment by justinmk** [comment, marked as answer]
Logging LHS/RHS is possible now by handling the `CmdAtom` event (Neovim 0.13 / nightly).

> shows what functions a user is calling, like capturing if they just called `vim.lsp.buf.definition`

That is not currently possible, and not really sure how it should work. Lua functions are often anonymous closures. We could optionally include a `traceback` (stacktrace) field in the `CmdAtom` event. Is that what you are looking for?
